File: app/src/main/java/com/johnpetitto/orachat/ui/chatroom/ChatroomPresenter.java
package com.johnpetitto.orachat.ui.chatroom;

import com.johnpetitto.orachat.data.chat.ChatModel;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;

public class ChatroomPresenter {
    private ChatroomView view;
    private ChatModel model;
    private CompositeDisposable disposables = new CompositeDisposable();

    public ChatroomPresenter(ChatroomView view, ChatModel model) {
        this.view = view;
        this.model = model;
    }

    public void onDestroy() {
        disposables.dispose();
    }

    public void createNewChat(String name, String message) {
        disposables.add(model.createNewChat(name, message)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(chatMessage -> view.showNewMessage(chatMessage),
                        throwable -> view.showError())
        );
    }

    public void getInitialMessages(long chatId) {
        view.showLoading(true);

        disposables.add(model.getMessagesForChat(chatId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(chatMessages -> {
                    view.showInitialMessages(chatMessages);
                    view.showLoading(false);
                }, throwable -> {
                    view.showError();
                    view.showLoading(false);
                })
        );
    }

    public void getMoreMessages() {
        disposables.add(model.loadMoreMessages()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(messages -> {
                    view.showMoreMessages(messages);
                    view.pageLoaded();
                }, throwable -> view.pageLoaded(), () -> view.pageLoaded())
        );
    }

    public void sendMessage(String message) {
        disposables.add(model.sendMessage(message)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(chatMessage -> view.showNewMessage(chatMessage),
                        throwable -> view.showError())
        );
    }

    public void updateChatName(String name) {
        model.updateChatName(name)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnError(throwable -> view.showError())
                .subscribe();
    }
}

File: app/src/main/java/com/johnpetitto/orachat/ui/chatroom/ChatroomPresenter.kt
package com.johnpetitto.orachat.ui.chatroom

import com.johnpetitto.orachat.data.chat.ChatModel
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable

class ChatroomPresenter(private val view: ChatroomView, private val model: ChatModel) {
    private var disposables = CompositeDisposable()

    fun destroy() = disposables.dispose()

    fun createNewChat(name: String, message: String) {
        disposables.add(model.createNewChat(name, message)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { view.showNewMessage(it) },
                        { view.showError() }
                )
        )
    }

    fun getInitialMessages(chatId: Long) {
        view.showLoading(true)

        disposables.add(model.getMessagesForChat(chatId)
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate { view.showLoading(false) }
                .subscribe(
                        { view.showInitialMessages(it) },
                        { view.showError() }
                )
        )
    }

    fun getMoreMessages() {
        disposables.add(model.loadMoreMessages()
                .onErrorComplete()
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate { view.pageLoaded() }
                .subscribe { view.showMoreMessages(it) }
        )
    }

    fun sendMessage(message: String) {
        disposables.add(model.sendMessage(message)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { view.showNewMessage(it) },
                        { view.showError() }
                )
        )
    }

    fun updateChatName(name: String) {
        disposables.add(model.updateChatName(name)
                .observeOn(AndroidSchedulers.mainThread())
                .doOnError { view.showError() }
                .subscribe()
        )
    }
}
